### DISCUSSION TITLE
Benchmark fails to compile with Xcode 16

### DISCUSSION THREAD

**Discussion by BjornRuud** [pointfreeco/swift-parsing#363] [General]
The benchmark will not compile with Xcode 16, it fails in JSON.swift (latest on main):

https://github.com/pointfreeco/swift-parsing/blob/82f10fc30e764f6ed680f9574ba0ac94743d8643/Sources/swift-parsing-benchmark/JSON.swift#L71

> Failed to produce diagnostic for expression; please submit a bug report (https://swift.org/contributing/#reporting-bugs)

**Comment by DivineDominion** [comment]
I got a similar error in a production app where Xcode 15.1/Swift 5.9.2 can infer the type just fine, while Xcode 15.4/Swift 5.10 can't.

I wonder if there's a way to flatten the parser builder expressions?



The inferred type is this in our case:

`Many<Substring.UTF8View, ParserBuilder<Substring.UTF8View>.SkipSecond<ParserBuilder<Substring.UTF8View>.Take2<ParserBuilder<Substring.UTF8View>.SkipSecond<ParserBuilder<Substring.UTF8View>.SkipSecond<ParserBuilder<Substring.UTF8View>.SkipSecond<ParserBuilder<Substring.UTF8View>.SkipFirst<Parsers.OverridePrinting<Whitespace<PartialRangeFrom<Int>, Conversions.Identity<Substring.UTF8View>>, EmptyParserPrinter>, Parsers.OverridePrinting<JSONStringParser, JSONStringParser>>, Parsers.OverridePrinting<Whitespace<PartialRangeFrom<Int>, Conversions.Identity<Substring.UTF8View>>, EmptyParserPrinter>>, Parsers.OverridePrinting<String.UTF8View, String.UTF8View>>, Parsers.OverridePrinting<Whitespace<PartialRangeFrom<Int>, Conversions.Identity<Substring.UTF8View>>, EmptyParserPrinter>>, JSONScalarParser>, Parsers.OverridePrinting<Whitespace<PartialRangeFrom<Int>, Conversions.Identity<Substring.UTF8View>>, EmptyParserPrinter>>, JSONFlatObject, String.UTF8View, Parse<Substring.UTF8View, ParserBuilder<Substring.UTF8View>.SkipSecond<ParserBuilder<Substring.UTF8View>.SkipSecond<Parsers.OverridePrinting<OneOf<Substring.UTF8View, (), OneOfBuilder<Substring.UTF8View, ()>.OneOf2<String.UTF8View, Whitespace<PartialRangeFrom<Int>, Conversions.Identity<Substring.UTF8View>>>>, EmptyParserPrinter>, Parsers.OverridePrinting<Whitespace<PartialRangeFrom<Int>, Conversions.Identity<Substring.UTF8View>>, EmptyParserPrinter>>, Parsers.OverridePrinting<String.UTF8View, String.UTF8View>>>, ()>`

From this expression:

```swift
        Many(0..., into: JSONFlatObject()) { object, pair in
            let (name, value) = pair
            object[name] = value
        } decumulator: { object in
            object.keys
                .map { key -> (String, JSONScalar) in
                    return ("\(key)", object[key]!)
                }
                .reversed()
                .makeIterator()
        } element: {
            Whitespace(0...).noPrinting()
            JSONStringParser()
                .prettyPrinter(printOptions) {
                    JSONStringParser(printOptions: printOptions.indented())
                }
            Whitespace(0...).noPrinting()
            ":".utf8
                .prettyPrinter(printOptions) { ": ".utf8 }
            Whitespace(0...).noPrinting()
            JSONScalarParser()
            Whitespace(0...).noPrinting()
        } separator: {
            ",".utf8
        } terminator: {
            Parse {
                OneOf {
                    ",".utf8
                    Whitespace(0...)
                }
                .noPrinting()

                Whitespace(0...).noPrinting()

                "}".utf8
                    .prettyPrinter(printOptions) {
                        ("\n" + printOptions.indentation + "}").utf8
                    }
            }
        }
```

**Comment by mbrandonw** [comment]
Hi @DivineDominion and @BjornRuud, as far as I can tell this was fixed in #355. We just needed to add a type annotation to get it compiling again, and it now compiles just fine for me in 16.0 and 16.1.

Do you have an example of a parser that does not compile in Xcode 16?

> I wonder if there's a way to flatten the parser builder expressions?

Just as with SwiftUI, you can break these parsers into smaller ones. For example, the `element` and `terminator` parsers are quite complex, so you can extract them into a dedicated `Parser` conformance.

I am going to convert this to a discussion since I am not sure it is an issue anymore. Let's continue the conversation over there.

**Reply by BjornRuud** [reply]
Yes benchmark on main compiles for me in Xcode 16.1. 👍

**Reply by DivineDominion** [reply]
Helping the compiler with type annotations fixed it in the library (and also in my app's parser example) 👍

I only noticed this very late after reinstalling Xcode 15.1 to get the inferred type (which also works when added as an `as` type hint).

I'd like to suggest a recommendation in the docstring to annotate the types in the accumulator block, or in the `Many` type docs, then it's visible from browsing the documentation page.
